File: library/src/main/java/me/proxer/library/api/user/UserTopTenEndpoint.java
package me.proxer.library.api.user;

import lombok.Setter;
import lombok.experimental.Accessors;
import me.proxer.library.api.Endpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entity.user.TopTenEntry;
import me.proxer.library.enums.Category;

import javax.annotation.Nullable;
import java.util.List;

/**
 * Endpoint for requesting an user's top ten list.
 *
 * @author Ruben Gees
 */
@Accessors(fluent = true)
public final class UserTopTenEndpoint implements Endpoint<List<TopTenEntry>> {

    private final InternalApi internalApi;

    @Nullable
    private final String userId;

    @Nullable
    private final String username;

    /**
     * Sets the category to filter by.
     */
    @Nullable
    @Setter
    private Category category;

    /**
     * Sets if hentai should be included in the result.
     */
    @Nullable
    @Setter
    private Boolean includeHentai;

    UserTopTenEndpoint(final InternalApi internalApi, @Nullable final String userId,
                       @Nullable final String username) {
        if (userId == null && username == null) {
            throw new IllegalArgumentException("You must pass either an userId or an username.");
        }

        this.internalApi = internalApi;
        this.userId = userId;
        this.username = username;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public ProxerCall<List<TopTenEntry>> build() {
        return internalApi.topTen(userId, username, category, includeHentai);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/user/UserTopTenEndpoint.kt
package me.proxer.library.api.user

import me.proxer.library.ProxerCall
import me.proxer.library.api.Endpoint
import me.proxer.library.entity.user.TopTenEntry
import me.proxer.library.enums.Category

/**
 * Endpoint for requesting an user's top ten list.
 *
 * @author Ruben Gees
 */
class UserTopTenEndpoint internal constructor(
    private val internalApi: InternalApi,
    private val userId: String?,
    private val username: String?
) : Endpoint<List<TopTenEntry>> {

    private var category: Category? = null
    private var includeHentai: Boolean? = null

    init {
        require(userId.isNullOrBlank().not() || username.isNullOrBlank().not()) {
            "You must pass either an userId or an username."
        }
    }

    /**
     * Sets the category to filter by.
     */
    fun category(category: Category?) = this.apply { this.category = category }

    /**
     * Sets if hentai should be included in the result.
     */
    fun includeHentai(includeHentai: Boolean? = true) = this.apply { this.includeHentai = includeHentai }

    override fun build(): ProxerCall<List<TopTenEntry>> {
        return internalApi.topTen(userId, username, category, includeHentai)
    }
}
